pkg/operator/apis/monitoring/v1alpha1/types.go

```diff
@@ -402,25 +429,132 @@ func (pm *PodMonitoring) endpontScrapeConfig(index int) (*promconfig.ScrapeConfi
+		return nil, errors.Wrap(err, "invalid scrape configuration")
+	}
+	return scrapeCfg, nil
+}
+
+// convertRelabelingRule converts the rule to a relabel configuration. An error is returned
+// if the rule would modify one of the protected labels.
+func convertRelabelingRule(r RelabelingRule) (*relabel.Config, error) {
+	rcfg := &relabel.Config{
+		// Upstream applies ToLower when digesting the config, so we allow the same.
+		Action:      relabel.Action(strings.ToLower(r.Action)),
+		TargetLabel: r.TargetLabel,
+		Separator:   r.Separator,
+		Replacement: r.Replacement,
+		Modulus:     r.Modulus,
+	}
+	for _, n := range r.SourceLabels {
+		rcfg.SourceLabels = append(rcfg.SourceLabels, prommodel.LabelName(n))
+	}
+	// We must only set the regex if its not empty. Like in other cases, the Prometheus code does
+	// not setup the structs correctly and this would default to the string "null" when marshalled,
+	// which is then interpreted as a regex again when read by Prometheus.
+	if r.Regex != "" {
+		re, err := relabel.NewRegexp(r.Regex)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid regex %q", r.Regex)
+		}
+		rcfg.Regex = re
+	}
+
+	// Validate that the protected target labels are not mutated by the provided relabeling rules.
+	switch rcfg.Action {
+	case relabel.Replace, relabel.HashMod:
+		// These actions write into the target label and it must not be a protected one.
+		if isProtectedLabel(r.TargetLabel) {
+			return nil, errors.Errorf("cannot relabel with action %q onto protected label %q", r.Action, r.TargetLabel)
+		}
+	case relabel.LabelDrop:
+		if matchesAnyProtectedLabel(rcfg.Regex) {
+			return nil, errors.Errorf("regex %s would drop at least one of the protected labels %s", r.Regex, strings.Join(protectedLabels, ", "))
+		}
+	case relabel.LabelKeep:
+		// Keep drops all labels that don't match the regex. So all protected labels must
+		// match keep.
+		if !matchesAllProtectedLabels(rcfg.Regex) {
+			return nil, errors.Errorf("regex %s would drop at least one of the protected labels %s", r.Regex, strings.Join(protectedLabels, ", "))
+		}
+	case relabel.LabelMap:
+		// It is difficult to prove for certain that labelmap does not override a protected label.
+		// Thus we just prohibit its use for now.
+		// The most feasible way to support this would probably be store all protected labels
+		// in __tmp_protected_<name> via a replace rule, then apply labelmap, then replace the
+		// __tmp label back onto the protected label.
+		return nil, errors.Errorf("relabeling with action %q not allowed", r.Action)
+	case relabel.Keep, relabel.Drop:
+		// These actions don't modify a series and are OK.
+	default:
+		return nil, errors.Errorf("unknown relabeling action %q", r.Action)
+	}
+	return rcfg, nil
+}
+
+var protectedLabels = []string{
+	export.KeyProjectID,
+	export.KeyLocation,
+	export.KeyCluster,
+	export.KeyNamespace,
+	export.KeyJob,
+	export.KeyInstance,
+	"__address__",
```

Comment:
Why do we need to protect the __address__ label? Does upstream not protect it during unmarshalling or reading?

Reply:
Upstream actually deliberate makes the scraped IP address available under that label so one can modify it via relabeling. We just don't want to allow it. It's no longer available during metric relabeling, which this PR adds, but in principle we also need to protect it in `targetLabels.fromPod`.

Reply:
Can we at least at a TODO for that, probably in `labelMappingRelabelConfigs`?

Reply:
A TODO about what?

Reply:
Ah yes - ignore me, I got confused